File: src/main/java/com/core/base/corebase/domain/user/Member.java
package com.core.base.corebase.domain.user;

import com.core.base.corebase.domain.organization.Organization;
import com.core.base.corebase.domain.organization.Team;
import com.core.base.corebase.domain.user.code.MemberState;
import com.core.base.corebase.domain.user.code.PermissionType;
import com.core.base.corebase.domain.user.code.UserState;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

@Entity
@Table(name = "member")
@Getter
@Setter
public class Member {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String email;

    @Column(nullable = false)
    private String permission;

    @Column(nullable = false)
    private String state;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "team_id")
    private Team team;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "organization_id", nullable = false)
    private Organization organization;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "uid")
    private User user;

    protected Member() {
        // For JPA
    }

    public Member(String email, String name, User user,
                 Organization organization, Team team,
                 PermissionType permission, MemberState state) {
        this.email = email;
        this.name = name;
        this.user = user;
        this.organization = organization;
        this.team = team;
        this.permission = permission.name();
        this.state = state != null ? state.name() : MemberState.WAIT.name();
    }

    public boolean isWait() {
        return MemberState.WAIT.equals(getState());
    }

    public void updateJoin(User user) {
        this.user = user;
        this.state = MemberState.JOIN.name();
    }

    public void updateUser(User user) {
        this.user = user;
    }

    public MemberState getState() {
        if(state == null) {
            return null;
        }
        return MemberState.valueOf(state);
    }

    public PermissionType getPermission() {
        if(permission == null) {
            return null;
        }
        return PermissionType.valueOf(permission);
    }
}

File: src/main/kotlin/com/core/base/corebase/domain/user/Member.kt
package com.core.base.corebase.domain.user

import com.core.base.corebase.domain.organization.Organization
import com.core.base.corebase.domain.organization.Team
import com.core.base.corebase.domain.user.code.MemberState
import com.core.base.corebase.domain.user.code.PermissionType
import jakarta.persistence.*

@Entity( name = "member")
class Member(
    email: String,
    name: String,
    user: User?, //User
    organization : Organization, //Organization
    team: Team?, //Team
    permission: PermissionType,
    state: MemberState = MemberState.WAIT
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0L

    var name = name; protected set
    var email = email; protected set
    var permission = permission; protected set

    @Enumerated(EnumType.STRING)
    var state = state; protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "team_id", nullable = true)
    var team: Team? = team; protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "organization_id", nullable = false)
    var organization: Organization = organization; protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "uid", nullable = true)
    var user: User? = user; protected set


    fun isWait(): Boolean = state.equals(MemberState.WAIT)
    fun updateJoin(user: User) {
        this.user = user
        this.state = MemberState.JOIN
    }
}
